File: klinara-ios/klinara-ios/Services/Reports/ReportModels.swift
import Foundation

// Kaynak: `apps/api/src/modules/reporting/dto/report.dto.ts` ve
// `packages/shared/src/reports-api.ts`.
//
// **Tarih aralıkları yarı açıktır: `[from, to)`.** Ay raporunda `to` ayın son
// günü değil, ertesi ayın ilk günüdür.
//
// ⚠️ HİÇBİR SAYI BURADA HESAPLANMIYOR. Oranlar, toplamlar ve yüzde değişimler
// sunucudan geldiği gibi taşınıyor — Faz 5 ve 6'daki "bakiye istemcide
// hesaplanmaz" kuralının aynısı. Yerelde ikinci bir toplam tutmak, senkron
// kalması gereken üçüncü bir gerçek kaynağı yaratırdı.

// MARK: - Ortak

/// Yanıtın kapsamı.
///
/// `own`, sunucunun çağıranı kendi personel satırına KİLİTLEDİĞİ anlamına
/// gelir. İstemci bunu izin listesine bakıp tahmin ETMİYOR; sunucudan okuyor.
/// İki taraf kuralı ayrı yorumlasaydı, sunucu daraltırken uygulama "tüm klinik"
/// diye başlık atan bir rapor gösterebilirdi.
nonisolated enum ReportScopeKind: String, Decodable, Sendable {
    case all
    case own
}

nonisolated struct ReportPeriod: Decodable, Sendable, Equatable {
    let from: String
    /// HARİÇ.
    let to: String
}

/// Yüzde değişim. `nil` **kıyaslanamaz** demek (önceki dönem sıfır), `0` değil.
typealias ReportDelta = [String: Double?]

// MARK: - Doluluk

nonisolated enum OccupancyGrouping: String, Sendable, CaseIterable, Identifiable {
    case staff
    case branch
    case day

    var id: String { rawValue }

    var turkishName: String {
        switch self {
        case .staff: return "Personel"
        case .branch: return "Şube"
        case .day: return "Gün"
        }
    }
}

nonisolated struct OccupancyTotals: Decodable, Sendable, Equatable {
    let bookedMinutes: Int
    let availableMinutes: Int
    /// Yüzde. Mesai dışı randevu varsa 100'ü **aşabilir**.
    let occupancyRate: Double
}

nonisolated struct OccupancyRow: Decodable, Sendable, Identifiable, Equatable {
    /// Gün kırılımında `nil`: yerel tarih bir kimlik değil, etiketin kendisi.
    let groupId: String?
    let groupLabel: String
    let bookedMinutes: Int
    let availableMinutes: Int
    let occupancyRate: Double

    var id: String { groupId ?? groupLabel }
}

nonisolated struct OccupancyReport: Decodable, Sendable, Equatable {
    let scope: ReportScopeKind
    let period: ReportPeriod
    let totals: OccupancyTotals
    let data: [OccupancyRow]
    let previous: OccupancyTotals?
    let delta: ReportDelta?
}

// MARK: - Ciro

nonisolated enum RevenueGrouping: String, Sendable, CaseIterable, Identifiable {
    case service
    case package
    case staff
    case branch
    case day
    case method

    var id: String { rawValue }

    var turkishName: String {
        switch self {
        case .service: return "Hizmet"
        case .package: return "Paket"
        case .staff: return "Personel"
        case .branch: return "Şube"
        case .day: return "Gün"
        case .method: return "Ödeme yöntemi"
        }
    }
}

nonisolated struct RevenueTotals: Decodable, Sendable, Equatable {
    /// Pencerede AÇILAN ücret kalemleri (kuruş).
    let accruedMinor: Int
    /// Pencerede YAPILAN, iptal edilmemiş tahsilatlar (kuruş).
    let collectedMinor: Int
    let refundedMinor: Int
    let currency: String
}

nonisolated struct RevenueRow: Decodable, Sendable, Identifiable, Equatable {
    let groupId: String?
    let groupLabel: String
    let accruedMinor: Int
    let collectedMinor: Int

    var id: String { groupId ?? groupLabel }
}

/// ⚠️ `data` satırlarının tahsilat toplamı `totals.collectedMinor`DAN küçük
/// olabilir ve bu bir hata değil: eski bir borca bu dönemde yapılan tahsilatın
/// bağlanacağı kalem pencerede değildir. Ekran toplamı `totals`tan okuyor,
/// satırları toplayarak değil.
nonisolated struct RevenueReport: Decodable, Sendable, Equatable {
    let scope: ReportScopeKind
    let period: ReportPeriod
    let totals: RevenueTotals
    let data: [RevenueRow]
    let previous: RevenueTotals?
    let delta: ReportDelta?
}

// MARK: - Personel performansı

nonisolated struct StaffPerformanceRow: Decodable, Sendable, Identifiable, Equatable {
    let staffProfileId: String
    let staffName: String
    let completedServices: Int
    let revenueMinor: Int
    /// Ters kayıtlar düşülmüş NET tahakkuk.
    let commissionMinor: Int
    let bookedMinutes: Int
    let availableMinutes: Int
    let occupancyRate: Double

    var id: String { staffProfileId }
}

nonisolated struct StaffPerformanceReport: Decodable, Sendable, Equatable {
    let scope: ReportScopeKind
    let period: ReportPeriod
    let data: [StaffPerformanceRow]
    let currency: String
}

// MARK: - Gelmeme ve iptal

nonisolated enum NoShowGrouping: String, Sendable, CaseIterable, Identifiable {
    case staff
    case branch
    case service
    case day

    var id: String { rawValue }

    var turkishName: String {
        switch self {
        case .staff: return "Personel"
        case .branch: return "Şube"
        case .service: return "Hizmet"
        case .day: return "Gün"
        }
    }
}

nonisolated struct NoShowTotals: Decodable, Sendable, Equatable {
    let total: Int
    let completed: Int
    let noShow: Int
    let cancelled: Int
    let noShowRate: Double
    let cancellationRate: Double
}

nonisolated struct NoShowRow: Decodable, Sendable, Identifiable, Equatable {
    let groupId: String?
    let groupLabel: String
    let total: Int
    let completed: Int
    let noShow: Int
    let cancelled: Int
    let noShowRate: Double
    let cancellationRate: Double

    var id: String { groupId ?? groupLabel }
}

nonisolated struct NoShowByOrigin: Decodable, Sendable, Identifiable, Equatable {
    let origin: String
    let total: Int
    let completed: Int
    let noShow: Int
    let cancelled: Int
    let noShowRate: Double
    let cancellationRate: Double

    var id: String { origin }

    var turkishName: String { origin == "online" ? "Online" : "Klinikten" }
}

nonisolated struct NoShowReport: Decodable, Sendable, Equatable {
    let period: ReportPeriod
    let totals: NoShowTotals
    let data: [NoShowRow]
    let byOrigin: [NoShowByOrigin]
    let previous: NoShowTotals?
    let delta: ReportDelta?
}

// MARK: - Kazanım ve geri dönüş

nonisolated struct AcquisitionRow: Decodable, Sendable, Identifiable, Equatable {
    /// `customers.source`; girilmemişse `nil`.
    let source: String?
    let customers: Int

    var id: String { source ?? "__unknown__" }

    var turkishName: String { source ?? "Belirtilmemiş" }
}

nonisolated struct RetentionTotals: Decodable, Sendable, Equatable {
    /// Penceredeki İLK tamamlanmış randevusu olan müşteriler.
    let newCustomers: Int
    let returningCustomers: Int
    let activeCustomers: Int
    let returningRate: Double
}

nonisolated struct CohortReturn: Decodable, Sendable, Identifiable, Equatable {
    let withinDays: Int
    let returned: Int
    let rate: Double

    var id: Int { withinDays }
}

/// ⚠️ `cohorts` oranları dönem bugüne yakınsa yapısal olarak düşük çıkar:
/// müşterilerin 90 günü henüz dolmamıştır. Sunucu bunu "düzeltmiyor" (kohortu
/// kırpmak sayının anlamını gizlerdi); ekran uyarıyı gösteriyor.
nonisolated struct RetentionReport: Decodable, Sendable, Equatable {
    let period: ReportPeriod
    let totals: RetentionTotals
    let acquisition: [AcquisitionRow]
    let cohorts: [CohortReturn]
    let previous: RetentionTotals?
    let delta: ReportDelta?
}

File: klinara-android/app/src/main/java/com/klinara/android/services/reports/ReportModels.kt
package com.klinara.android.services.reports

import com.klinara.android.services.crm.CustomerSource
import com.klinara.android.services.networking.PageInfo
import com.klinara.android.services.networking.WireEnumSerializer
import kotlinx.serialization.Serializable

// Kaynak: `apps/api/src/modules/reporting/dto/report.dto.ts`, `packages/shared/src/reports-api.ts`;
// iOS `ReportModels.swift` paritesi.
//
// ⚠️ HİÇBİR SAYI BURADA HESAPLANMIYOR. Oranlar, toplamlar ve yüzde değişimler sunucudan geldiği
// gibi taşınıyor (A5/A6'daki "bakiye istemcide hesaplanmaz" kuralı). Para `Long` kuruş; oranlar
// sunucunun iki haneye yuvarladığı yüzde (`Double` yalnız GÖSTERİM için — hesap yok).

/** Yanıtın kapsamı — istemci izinden TÜRETMEZ, sunucunun dediğini gösterir. */
@Serializable(with = ReportScopeSerializer::class)
enum class ReportScope(
    val wire: String,
) {
    All("all"),
    Own("own"),
    Unknown("unknown"),
    ;

    companion object {
        fun from(wire: String): ReportScope = entries.firstOrNull { it.wire == wire } ?: Unknown
    }
}

internal object ReportScopeSerializer : WireEnumSerializer<ReportScope>("ReportScope", ReportScope::from, { it.wire })

/** Sunucunun geri yansıttığı aralık — gönderildiği gibi. */
@Serializable
data class ReportPeriodEcho(
    val from: String,
    val to: String,
)

/**
 * Yüzde değişim: anahtar alan adı, değer göreli % (iki hane). **`null` ≠ `0`**: `null`
 * "karşılaştırılamaz" (önceki dönem 0, bu dönem değil); `0` yalnız ikisi de 0 iken gelir.
 * Anahtar kümesi rapora göre değişir — hiçbir anahtar varsayılmaz.
 */
typealias ReportDelta = Map<String, Double?>

/** Rapor sorgusu — sayfalamadan bağımsız kısım. */
data class ReportQuery(
    val period: ReportPeriod,
    /** `null` → sunucu "erişebildiğim şubeler"i kullanır (kiracının tamamı DEĞİL). */
    val branchId: String?,
    val compareToPrevious: Boolean = false,
)

/** Sayfalama isteğe bağlı: [limit] yoksa sunucu TÜM satırları döndürür. */
data class ReportPage(
    val limit: Int? = null,
    val cursor: String? = null,
) {
    companion object {
        val UNPAGED = ReportPage()
    }
}

/** Gruplamalar — sunucunun `*_GROUPINGS` kümeleri, Türkçe adlarla. */
enum class OccupancyGrouping(
    val wire: String,
    val turkishName: String,
) {
    Staff("staff", "Personel"),
    Branch("branch", "Şube"),
    Day("day", "Gün"),
}

enum class RevenueGrouping(
    val wire: String,
    val turkishName: String,
) {
    Service("service", "Hizmet"),
    Package("package", "Paket"),
    Staff("staff", "Personel"),
    Branch("branch", "Şube"),
    Day("day", "Gün"),
    Method("method", "Ödeme yöntemi"),
}

enum class NoShowGrouping(
    val wire: String,
    val turkishName: String,
) {
    Staff("staff", "Personel"),
    Branch("branch", "Şube"),
    Service("service", "Hizmet"),
    Day("day", "Gün"),
}

// --- Doluluk ---

@Serializable
data class OccupancyTotals(
    val bookedMinutes: Int = 0,
    val availableMinutes: Int = 0,
    /** Yüzde; mesai dışı randevu varsa **100'ü aşabilir**. */
    val occupancyRate: Double = 0.0,
)

@Serializable
data class OccupancyRow(
    /** Gün gruplamasında `null`; etiket o zaman `YYYY-MM-DD`. */
    val groupId: String? = null,
    val groupLabel: String,
    val bookedMinutes: Int = 0,
    val availableMinutes: Int = 0,
    val occupancyRate: Double = 0.0,
) {
    val id: String get() = groupId ?: groupLabel
}

@Serializable
data class OccupancyReport(
    val scope: ReportScope = ReportScope.All,
    val period: ReportPeriodEcho,
    val totals: OccupancyTotals,
    val data: List<OccupancyRow> = emptyList(),
    val pageInfo: PageInfo? = null,
    val previous: OccupancyTotals? = null,
    val delta: ReportDelta? = null,
)

// --- Ciro ---

@Serializable
data class RevenueTotals(
    val accruedMinor: Long = 0,
    val collectedMinor: Long = 0,
    val refundedMinor: Long = 0,
    val currency: String = "TRY",
)

@Serializable
data class RevenueRow(
    val groupId: String? = null,
    /** Atıfsız satırda `"—"`; yöntem kırılımında ham yöntem (`card`, `cash`…). */
    val groupLabel: String,
    /** Yöntem kırılımında HER ZAMAN 0: ödeme yöntemi bir tahsilat özelliği, kalem değil. */
    val accruedMinor: Long = 0,
    val collectedMinor: Long = 0,
) {
    val id: String get() = groupId ?: groupLabel
}

@Serializable
data class RevenueReport(
    val scope: ReportScope = ReportScope.All,
    val period: ReportPeriodEcho,
    val totals: RevenueTotals,
    val data: List<RevenueRow> = emptyList(),
    val pageInfo: PageInfo? = null,
    val previous: RevenueTotals? = null,
    val delta: ReportDelta? = null,
) {
    /**
     * Toplamlarda hareket var mı? iOS toplam kartını `data` boşken gizliyordu; oysa eski bir
     * borca bu dönemde yapılan tahsilat satırsız ama toplamlı bir dönem üretir.
     */
    val hasMovement: Boolean
        get() =
            data.isNotEmpty() ||
                listOf(totals.accruedMinor, totals.collectedMinor, totals.refundedMinor).any { it != 0L }
}

/** Ödeme yöntemi etiketleri — sunucunun `PaymentMethod` kümesi; bilinmeyen HAM kalır. */
object PaymentMethodLabels {
    private val LABELS =
        mapOf(
            "cash" to "Nakit",
            "card" to "Kart",
            "bank_transfer" to "Havale / EFT",
            "gift_voucher" to "Hediye çeki",
            "other" to "Diğer",
        )

    fun label(wire: String): String = LABELS[wire] ?: wire
}

// --- Personel performansı ---

@Serializable
data class StaffPerformanceRow(
    val staffProfileId: String,
    val staffName: String,
    val completedServices: Int = 0,
    val revenueMinor: Long = 0,
    /** Ters kayıtlar düşülmüş. */
    val commissionMinor: Long = 0,
    val bookedMinutes: Int = 0,
    val availableMinutes: Int = 0,
    val occupancyRate: Double = 0.0,
)

/** `totals`/`previous`/`delta` YOK — sunucu `compareTo`'yu bu raporda yok sayıyor. */
@Serializable
data class StaffPerformanceReport(
    val scope: ReportScope = ReportScope.All,
    val period: ReportPeriodEcho,
    val data: List<StaffPerformanceRow> = emptyList(),
    val pageInfo: PageInfo? = null,
    val currency: String = "TRY",
)

// --- Gelmeme ve iptal ---

@Serializable
data class NoShowTotals(
    val total: Int = 0,
    val completed: Int = 0,
    val noShow: Int = 0,
    val cancelled: Int = 0,
    val noShowRate: Double = 0.0,
    val cancellationRate: Double = 0.0,
)

@Serializable
data class NoShowRow(
    val groupId: String? = null,
    val groupLabel: String,
    val total: Int = 0,
    val completed: Int = 0,
    val noShow: Int = 0,
    val cancelled: Int = 0,
    val noShowRate: Double = 0.0,
    val cancellationRate: Double = 0.0,
) {
    val id: String get() = groupId ?: groupLabel
}

@Serializable
data class NoShowByOrigin(
    val origin: String,
    val total: Int = 0,
    val completed: Int = 0,
    val noShow: Int = 0,
    val cancelled: Int = 0,
    val noShowRate: Double = 0.0,
    val cancellationRate: Double = 0.0,
) {
    val turkishName: String get() = if (origin == "online") "Online" else "Klinikten"
}

/** `scope` alanı YOK — bu uç `appointment:read.all` istiyor, kendi kapsamı olmuyor. */
@Serializable
data class NoShowReport(
    val period: ReportPeriodEcho,
    val totals: NoShowTotals,
    val data: List<NoShowRow> = emptyList(),
    val pageInfo: PageInfo? = null,
    val byOrigin: List<NoShowByOrigin> = emptyList(),
    val previous: NoShowTotals? = null,
    val delta: ReportDelta? = null,
)

// --- Kazanım ve geri dönüş ---

@Serializable
data class RetentionTotals(
    val newCustomers: Int = 0,
    val returningCustomers: Int = 0,
    val activeCustomers: Int = 0,
    val returningRate: Double = 0.0,
)

@Serializable
data class AcquisitionRow(
    /** Müşteri kaynağı (`customers.source`); girilmemişse `null`. */
    val source: String? = null,
    val customers: Int = 0,
) {
    val id: String get() = source ?: UNKNOWN_ID

    /**
     * iOS ham değeri gösteriyordu (`instagram`); kaynak müşteri kartındaki enum'un aynısı, bu
     * yüzden kartla aynı Türkçe ad kullanılıyor. Tanınmayan değer ham kalır.
     */
    val turkishName: String
        get() {
            val raw = source ?: return "Belirtilmemiş"
            val known = CustomerSource.from(raw)
            return if (known == CustomerSource.Unknown) raw else known.turkishName
        }

    companion object {
        const val UNKNOWN_ID = "__unknown__"
    }
}

@Serializable
data class CohortReturn(
    val withinDays: Int,
    val returned: Int = 0,
    val rate: Double = 0.0,
)

/** Sayfasız; `scope` ve `pageInfo` YOK. Müşteri kimliği/telefonu hiç taşımaz. */
@Serializable
data class RetentionReport(
    val period: ReportPeriodEcho,
    val totals: RetentionTotals,
    val acquisition: List<AcquisitionRow> = emptyList(),
    val cohorts: List<CohortReturn> = emptyList(),
    val previous: RetentionTotals? = null,
    val delta: ReportDelta? = null,
)

/** Dışa aktarılabilir raporlar — sunucu yolları ve dosya adı öneki. */
enum class ReportKind(
    val path: String,
    val filePrefix: String,
) {
    Occupancy("occupancy", "doluluk"),
    Revenue("revenue", "ciro"),
    StaffPerformance("staff-performance", "personel-performans"),
    NoShow("no-show", "gelmeme"),
    Retention("retention", "kazanim"),
}
